src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+                                try:
+                                    self.outcome = json.loads(
+                                        event.additional_data.decode("utf-8")
+                                    )["reason"]
+                                except Exception:
+                                    self.outcome = str(event.additional_data[:100])
+                            else:
+                                self.outcome = str(event.error_code)
+                        logger.info("%s %s", self, self.outcome)
+                    elif not sid and error is not None:
+                        # FIXME break the connection,but give users a chance to complete
+                        logger.warning("Bad error %s", event)
+                        self.outcome = str(error)
+                        self.closing = True
+                    else:
+                        if isinstance(event, h2.events.DataReceived):
+                            # Stream flow control is responsibility of the channel.
+                            # Connection flow control is handled here.
+                            self.protocol.increment_flow_control_window(
+                                event.flow_controlled_length
+                            )
+                        if ch:
+                            ch.ev.append(event)
+                            if not ch.fut.done():
+                                ch.fut.set_result(None)
+
+                # Somewhat inefficient: wake up background writer just in case
+                # it could be that we've received something that h2 needs to acknowledge
+                self.should_write.set()
+
+                # FIXME notify users about possible change to `.blocked`
+                # FIXME selective:
+                # * h2.events.WindowUpdated
+                # * max_concurrent_streams change
+                # * [maybe] starting a stream
+                # * a stream getting closed (but not half-closed)
+                # * closing / closed change
+        except ConnectionError as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background read task died")
+        finally:
+            self.closing = self.closed = True
+            for sid, ch in self.channels.items():
+                if ch.fut and not ch.fut.done():
+                    ch.fut.set_exception(Closed(self.outcome))
+
+    async def post(self, req: "Request") -> "Response":
+        assert len(req.body) <= MAX_NOTIFICATION_PAYLOAD_SIZE
+
+        now = time()
+        if now > req.deadline:
+            raise Timeout("Request timed out")
+        if self.closing or self.closed:
+            raise Closed(self.outcome)
+        if self.blocked:
+            raise Blocked()
+
+        try:
+            sid = self.protocol.get_next_available_stream_id()
+        except h2.NoAvailableStreamIDError:
```

Comment:
`h2` has no attribute `NoAvailableStreamIDError` (at least in 3.2.0). TIL: the name passed to `except` is only checked in case there is an exception.

Reply:
hopefully done